### PR TITLE
DRIVERS-996: Versioned API

### PR DESCRIPTION
**Pull request by alcaeus**
DRIVERS-996

This PR contains the versioned API specification, as well as the changes necessary to test versioned API with the unified test runner.

### CODE REVIEW

File: source/versioned-api/versioned-api.rst [hunk 1]
```diff
@@ -0,0 +1,273 @@
...
+participate in executing operations. Therefore, when determining application
+compatibility with MongoDB, both the driver and the server behavior must be
+taken into account.
+
+An application can specify the server API version when creating MongoClient.
+When this is done:
+
+- The client sends the specified API version to the server, causing the server
+  to behave in a manner compatible with that API version.
+- The driver will behave in a manner compatible with the driver's behavior as of
```
Comment (patrickfreed) [inline source/versioned-api/versioned-api.rst:55]: I'm finding this line is a bit hard to parse, could it be reworded to be a bit more simple / straightforward?

File: source/versioned-api/versioned-api.rst [hunk 2]
```diff
@@ -0,0 +1,273 @@
...
+
+Drivers SHOULD group the ``serverApi`` option with other similar client options
+like ``autoEncryptionOpts``.
+
+
+ServerApiVersion enumeration
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+This specification and subsequent specifications will define the known API
+versions. Drivers SHOULD define an enumeration containing the known API
```
Comment (patrickfreed) [inline source/versioned-api/versioned-api.rst:99]: One issue with using an enum for this is that a lot of languages allow exhaustive `switch` statements over the various cases, meaning adding a new case constitutes a breaking change (Rust doesn't have this problem, by the way). If we intend to add support for new API versions in non-major releases, then we might want to include the addition requirement here that "drivers MUST be able to introduce new versions without breaking source compatibility" or something like that.

File: source/versioned-api/versioned-api.rst [hunk 3]
```diff
@@ -0,0 +1,273 @@
...
+**ServerApi field**   **Command option**
+``version``           ``apiVersion``
+``strict``            ``apiStrict``
+``deprecationErrors`` ``apiDeprecationErrors``
+===================== ========================
+
+If an API version was declared, drivers MUST add the ``apiVersion`` option to
+every command that is sent to a server. Drivers MUST add the ``apiStrict`` and
+``apiDeprecationErrors`` options if they were specified by the user, even when
+the specified value is equal to the server default. Drivers SHOULD NOT add these
```
Comment (patrickfreed) [inline source/versioned-api/versioned-api.rst:158]: can we make this "SHOULD NOT" a "MUST NOT"? Or is there a circumstance where it might be necessary to send a value the user didn't specify?

File: source/versioned-api/versioned-api.rst [hunk 4]
```diff
@@ -0,0 +1,273 @@
...
+   }
+
+   class ServerApi {
+       version: string|ServerApiVersion;
+       strict: Optional<Boolean>; // Default false
+       deprecationErrors: Optional<Boolean>; // Default false
+   }
+
+Drivers SHOULD group the ``serverApi`` option with other similar client options
+like ``autoEncryptionOpts``.
```
Comment (patrickfreed) [inline source/versioned-api/versioned-api.rst:92]: can we add an explicit "Drivers MUST NOT allow specification of any versioned API options via the connection string" (or something like that) here? I think the first time this requirement shows up in this document is in the rationale, so we should probably call it out more directly here.

File: source/versioned-api/versioned-api.rst [hunk 5]
```diff
@@ -0,0 +1,273 @@
...
+Hardcode supported versions in drivers
+--------------------------------------
+
+Since a new API version might require driver changes (e.g. to account for
+removed commands), we don't yet know what changes drivers must make for a future
+version. Until we do, we must prevent users from choosing any unknown API
+version.
+
+
+Backward Compatibility
```
Comment (patrickfreed) [inline source/versioned-api/versioned-api.rst:244]: Should we mention the related policy changes in the spec at all? I imagine in the future we'll add spec language for an "experimental" feature, so we'll need to have that term defined in the specs somewhere.

File: source/versioned-api/versioned-api.rst [hunk 6]
```diff
@@ -0,0 +1,273 @@
...
+  version.
+
+Presently there is no specification for how a driver must behave when a
+particular server API version is requested, or what driver operations are
+subject to API compatibility guarantees. Such requirements may be stipulated in
+subsequent specifications.
+
+This specification requires MongoClient to validate that it supports the
+specified server API version, if any, but does not define what such support
+means.
```
Comment (patrickfreed) [inline source/versioned-api/versioned-api.rst:66]: It might be helpful for new readers of this spec if this was more prescriptive about support for version 1. Maybe something like "Drivers that support MongoDB 4.4 automatically support API version 1, but this specification does not define what such support means for later API versions."

File: source/versioned-api/versioned-api.rst (hunk 1 above)
Comment (alcaeus) [reply source/versioned-api/versioned-api.rst:55]: What about this?
> The driver will behave in a manner compatible with the driver's original behavior at the time the specified API version was released.

File: source/versioned-api/versioned-api.rst (hunk 3 above)
Comment (alcaeus) [reply source/versioned-api/versioned-api.rst:158]: I used "SHOULD NOT" to allow drivers to use a `bool` for `strict` and `deprecationErrors` without tracking whether the user supplied it. If I change this to "MUST NOT", I'd also remove the default values for the above options from the `ServerApi` class to ensure that they are empty by default, which indicates that the driver must not add them.

File: source/versioned-api/versioned-api.rst (hunk 4 above)
Comment (alcaeus) [reply source/versioned-api/versioned-api.rst:92]: Yep, seems sensible to explicitly mention this here 👍

File: source/versioned-api/versioned-api.rst (hunk 5 above)
Comment (alcaeus) [reply source/versioned-api/versioned-api.rst:244]: Since the policy is not public, I wanted to avoid mentioning it here. I can add a note about it for all driver engineers that will be reading this.

File: source/versioned-api/versioned-api.rst (hunk 6 above)
Comment (alcaeus) [reply source/versioned-api/versioned-api.rst:66]: Technically you are correct - there are a few specialties though. With `count` being removed from API Version 1 before the release, we'll have to update the CRUD specification to change how `estimatedDocumentCount` works. Similarly, `distinct` is not in API Version 1, so the `distinct` helper would also need a change.

With these two issues open, I wouldn't yet add this chapter, but deal with this once we get to the DRIVERS tickets for the two cases mentioned above.

File: source/versioned-api/versioned-api.rst (hunk 2 above)
Comment (alcaeus) [reply source/versioned-api/versioned-api.rst:101]: Technically, drivers don't have to add compatibility for a new API Version in the current major release series, but can do so in a new major release if that's required. I've asked the leads for input on how their languages treat this to make an informed decision and will revisit this once I have some answers.

Comment (alcaeus) [reply source/versioned-api/versioned-api.rst:101]: After discussion we've come to the conclusion that it makes sense to let drivers handle this individually without prescribing a one-size-fits-all solution. For example, a new enum value could produce a warning in C/C++ (via `-Wswitch-enum` in gcc), which in turn can cause compiler errors if warnings are treated as errors. However, since it is the user's choice to treat this strictly it is up to them to take care of this and such a change would be allowed in a minor release. Swift is a different story, as users may want to work with the enumeration in a `switch` statement, so adding a new version could be a BC break.

With this in mind, I'd suggest adding the following wording:
> Drivers MUST ensure that adding new API versions to this enumeration does not result in backward compatibility breaks in non-major releases. This can be the case in languages that allow exhaustive ``switch`` statements (e.g. Swift). In these cases, drivers should either refrain from using enumerations or only introduce new API versions in major releases.

This ensures that readers of the spec are aware of the potential issue, but can decide how they want to deal with it: either by not using enumerations (as is done in the Swift driver in some instances), by choosing to not treat it as a BC breaking change (as the C/C++ driver might want to do), or by avoiding the issue altogether and releasing a new major version.

File: source/versioned-api/tests/README.rst [hunk 7]
```diff
@@ -0,0 +1,16 @@
+===================
+Versioned API Tests
+===================
+
+.. contents::
+
+----
+
+Notes
```
Comment (jyemin) [inline source/versioned-api/tests/README.rst:9]: We also must require that drivers test against a server running with

    --setParameter requireApiVersion=true

with auth enabled.  Otherwise, we won't be testing that drivers will work in this scenario.

File: source/versioned-api/versioned-api.rst (hunk 2 above)
Comment (jyemin) [reply source/versioned-api/versioned-api.rst:101]: I'm fine with the decision to provide flexibility, but I don't think we should mention the option of a major release just for a new API version, which likely otherwise wouldn't require any other driver changes except the addition of the enumeration.  Major releases are painful for users and require a lot of planning, and that's precisely what we are trying to avoid.

Comment (jyemin) [reply source/versioned-api/versioned-api.rst:101]: I would just remove the last sentence.

File: source/versioned-api/versioned-api.rst [hunk 8]
```diff
@@ -0,0 +1,277 @@
...
+
+
+ServerApi class
+~~~~~~~~~~~~~~~
+
+The ``ServerApi`` class stores an API version, along with flags that decide
+whether or not unknown or deprecated commands in the specified API version
+trigger a server-side error. A ``version`` MUST be specified when declaring an
+API version, while the ``strict`` and ``deprecationErrors`` options are both
+optional and default to ``false``. The ``ServerApi`` class is considered
```
Comment (jyemin) [inline source/versioned-api/versioned-api.rst:120]: > are both optional and default to ``false``

If it's optional then it defaults to `unset` or `not present`.  And it contradicts what is said below:

> Drivers SHOULD NOT add these options if the user did not specify a value.

Comment (jyemin) [reply source/versioned-api/versioned-api.rst:120]: Do you mean to say:

> are either optional, or non-optional and default to ``false``

?

File: source/versioned-api/versioned-api.rst [hunk 9]
```diff
@@ -0,0 +1,277 @@
...
+every command that is sent to a server. Drivers MUST add the ``apiStrict`` and
+``apiDeprecationErrors`` options if they were specified by the user, even when
+the specified value is equal to the server default. Drivers SHOULD NOT add these
+options if the user did not specify a value.
+
+
+Cursors
+~~~~~~~
+
+The ``getMore`` command does not accept API parameters; cursors inherit their
```
Comment (jyemin) [inline source/versioned-api/versioned-api.rst:167]: Consider an explicit note about `killCursors` accepting API parameters.  It's weird that one does and the other doesn't, so I think it's worth calling out.

**Review by jyemin — CHANGES_REQUESTED** [review]

File: source/versioned-api/versioned-api.rst (hunk 1 above)
Comment (patrickfreed) [reply source/versioned-api/versioned-api.rst:55]: Hm, I think it's still a bit vague, e.g. what does "compatible" or "original behavior" mean in this case? This requirement  _has_ to be vague though because there isn't anything special for drivers to do yet, as is pointed out below. I think the issue here is that the only requirement we can feasibly impose is that the driver is compatible with a _server_ using that API version, not with _other versions of the driver_ using that API version. For instance, version 5.0 of a driver configured with `serverApiVersion: 1` does not necessarily need to exhibit the same behavior of version 6.0 of a driver configured with `serverApiVersion: 1` (e.g if 6.0 removed the `distinct` helper or something), though both versions do need to be able to _function_ with a server configured with `serverApiVersion: 1`.

So what about something as simple as the following:

>The driver will behave in a manner compatible with a server configured with that API version, regardless of the server's actual release version.

It's pretty weak, but I think it's the only meaningful requirement we can impose for now until more requirements emerge.

File: source/versioned-api/versioned-api.rst (hunk 2 above)
Comment (patrickfreed) [reply source/versioned-api/versioned-api.rst:101]: Yeah, I agree. Updating the default version API should certainly require a major release (imo), but simply adding support  for a new one shouldn't require one. The rest of the wording lgtm.

File: source/versioned-api/versioned-api.rst (hunk 3 above)
Comment (patrickfreed) [reply source/versioned-api/versioned-api.rst:160]: I think unless we wish to deviate from the server's default values for these, we should just not send them if the user didn't specify them. This would be consistent with how we handle other options that we forward along to the server.

File: source/versioned-api/versioned-api.rst (hunk 5 above)
Comment (patrickfreed) [reply source/versioned-api/versioned-api.rst:248]: Yeah true. I think whenever we do add the first "experimental" API to the specifications we can just tackle this issue then.

File: source/versioned-api/versioned-api.rst (hunk 1 above)
Comment (p-mongo) [reply source/versioned-api/versioned-api.rst:55]: I support Patrick's suggested wording.

File: source/versioned-api/versioned-api.rst (hunk 3 above)
Comment (p-mongo) [reply source/versioned-api/versioned-api.rst:160]: On line 160 what does "these" refer to? strict & deprecation errors or all 3 options?

Comment (p-mongo) [reply source/versioned-api/versioned-api.rst:160]: I am not following the reason for should vs must but regarding this:

>  I'd also remove the default values for the above options from the ServerApi class to ensure that they are empty by default, which indicates that the driver must not add them.

We already require drivers not sending default values for `multi` and `upsert` in crud. Therefore I expect it should be possible for drivers to not send default values here.

**Review by p-mongo — COMMENTED** [review]
I read the PR and I support the open comment threads. LGTM with the changes in comments applied.

File: source/versioned-api/versioned-api.rst (hunk 1 above)
Comment (alcaeus) [reply source/versioned-api/versioned-api.rst:55]: Sounds good, I applied your suggested wording.

File: source/versioned-api/versioned-api.rst (hunk 2 above)
Comment (alcaeus) [reply source/versioned-api/versioned-api.rst:101]: Removed the last sentence as per your suggestions.

File: source/versioned-api/versioned-api.rst (hunk 3 above)
Comment (alcaeus) [reply source/versioned-api/versioned-api.rst:160]: Changed "SHOULD NOT" to "MUST NOT" and reworded the last sentence for clarity:
> Drivers MUST NOT add any API versioning options if the user did not specify them.

File: source/unified-test-format/unified-test-format.rst [hunk 10]
```diff
@@ -343,6 +348,11 @@ The structure of this object is as follows:
   "sharded" topology, test runners MUST accept any type of sharded cluster (i.e.
   "sharded" implies "sharded-replicaset", but not vice versa).
 
+- ``serverParameters``: Optional object of server parameters to check against.
+  To check server parameters, drivers send a ``{ getParameter: '*' }`` command
+  to the server using the internal MongoClient. The result SHOULD be cached to
+  avoid repeated calls. The result is then compared to the expected server
+  parameters as described in the `Evaluating Matches`_ section.
```
[the remark this answers is not in the document]
Comment (alcaeus) [reply source/unified-test-format/unified-test-format.rst:355]: That sounds sensible. I've changed the paragraph to suggest either fetching single parameters or all of them and treat missing parameters as not equal. I've added a sentence to require test runners to evaluate these conditions in the order of the list above (min/max version, topology, parameters) which should eliminate cases where parameters don't exist, but I figured that if we're doing an exact match on parameters, a missing parameter should always cause tests to be skipped.

File: source/versioned-api/versioned-api.rst (hunk 8 above)
Comment (alcaeus) [reply source/versioned-api/versioned-api.rst:120]: Removed any mention of a default value corresponding with the above change requiring drivers to not send apiStrict and apiDeprecationErrors unless they were explicitly specified.

File: source/versioned-api/versioned-api.rst (hunk 9 above)
Comment (alcaeus) [reply source/versioned-api/versioned-api.rst:167]: Added in a new paragraph:
> In contrast, the ``killCursors`` command accepts API parameters and drivers MUST include them as they would with all other commands.

File: source/versioned-api/tests/README.rst (hunk 7 above)
Comment (alcaeus) [reply source/versioned-api/tests/README.rst:9]: Yes, I'll add that.

While setting up the mongo-orchestration config locally I realised that this doesn't work completely yet: the server is started correctly, but the `mo` binary does not return. This is due to mongo-orchestration not sending an api version for any commands that it wants to run, thus hanging. This can't be resolved until PyMongo is updated to support the versioned API.

File: source/versioned-api/versioned-api.rst (hunk 2 above)
Comment (jyemin) [reply source/versioned-api/versioned-api.rst:100]: lgtm, but leaving for patrick to resolve

File: source/versioned-api/tests/README.rst (hunk 7 above)
Comment (jyemin) [reply source/versioned-api/tests/README.rst:9]: Oh, ouch.  For purposes of this PR, it's sufficient to just add the prose here.  We can work out orchestration issues separately.

Comment (alcaeus) [reply source/versioned-api/tests/README.rst:9]: Right. Added the following paragraph:
> Drivers MUST test against a server with the ``requireApiVersion`` parameter
> enabled that also requires authentication. Since API versioning options can't be
> specified using the connection string, drivers MUST rely on environment
> variables to receive the API version to use in tests. The ``apiStrict`` and
> ``apiDeprecationErrors`` options are not required for this test.

Comment (jyemin) [reply source/versioned-api/tests/README.rst:9]: This still leaves a small hole in test coverage, as we don't ensure that apiVersion is included for all auth mechanisms. But I think it's the best we can do right now.

**Review by jyemin — APPROVED** [review]
LGTM!

**Review by patrickfreed — APPROVED** [review]
LGTM, looks great!

**Review by p-mongo — APPROVED** [review]

File: source/versioned-api/versioned-api.rst [hunk 11]
```diff
@@ -0,0 +1,282 @@
+===================
+Unified Test Format
```
Comment (patrickfreed) [inline source/versioned-api/versioned-api.rst:2]: Oh, just noticed that this needs to be updated. Surprised we all missed the biggest text in the spec :smile:
